### DISCUSSION TITLE
Unable to migrate from RELEASE.2021-07-08T01-15-01Z to RELEASE.2021-11-03T03-36-36Z

### DISCUSSION THREAD

**Discussion by gughappriya** [minio/minio#14263] [Q&A]
## Expected Behavior
if we use sse enabled in both versions of minio , when data thats added using RELEASE.2021-07-08T01-15-01Z upgraded to RELEASE.2021-11-03T03-36-36Z . The data should be upgraded with no issues
```
Priyasekar@PriyaSekars-MacBook-Pro ibm-minio-operator % mc cp --insecure sse-test/bucket/testing-encrypt.txt  ./test-downloaded-from-minio-1010.txt
success
```

## Current Behavior
if we use sse enabled in both versions of minio , when data thats added using RELEASE.2021-07-08T01-15-01Z upgraded to RELEASE.2021-11-03T03-36-36Z , results in error
```
Priyasekar@PriyaSekars-MacBook-Pro ibm-minio-operator % mc cp --insecure sse-test/bucket/testing-encrypt.txt  ./test-downloaded-from-minio-1010.txt
mc: <ERROR> Failed to copy `https://localhost:9000/bucket/testing-encrypt.txt`. Server side encryption specified but KMS is not configured
```

## Steps to Reproduce (for bugs)
1. Add data to bucket using RELEASE.2021-07-08T01-15-01Z 
2. Upgrade minio version to RELEASE.2021-11-03T03-36-36Z
3. Using minio cp to copy the data added from previous version of minio 
4. It fails with the error 
```
mc: <ERROR> Failed to copy `https://localhost:9000/bucket/testing-encrypt.txt`. Server side encryption specified but KMS is not configured
```

**Detailed story :**
So to give the full story of what we tried lets call this old release(RELEASE.2021-07-08T01-15-01Z ) as X and the later release(RELEASE.2021-11-03T03-36-36Z) we tried as Y
**Scenario :** We added data using the X version of minio and then later upgraded to Y version. The data we added was not readable anymore
1. With SSE enabled we added data to buckets using X version of minio.
```
mc --insecure config host add sse-test https://localhost:9000/
mc ls --insecure sse-test

mc cp --insecure --encrypt sse-test/bucket/ ./testing.txt sse-test/bucket/testing-encrypt.txt
./testing.txt:                         21 B / 21 B

mc stat --insecure sse-test/bucket/testing-encrypt.txt
Name      : testing-encrypt.txt
Date      : 2021-12-13 17:53:35 CST
Size      : 21 B
ETag      : 3a5eba4d91f3dbbe047f24ce02cd2530
Type      : file
Metadata  :
  Content-Type: text/plain
Encrypted :
  X-Amz-Server-Side-Encryption: AES256

mc cat --insecure sse-test/bucket/testing-encrypt.txt
Hello Hello Hello!!!
```
2. When we upgraded to Y version of minio we got the following error on reading
`cause: com.amazonaws.services.s3.model.AmazonS3Exception: Server side encryption specified but KMS is not configured (Service: Amazon S3; Status Code: 501; Error Code: NotImplemented; Request ID: 16BC9A97099E0EE9; S3 Extended Request ID: 52bff882-da55-46c7-a8bb-a7cb76534d2d), S3 Extended Request ID: 52bff882-da55-46c7-a8bb-a7cb76534d2d)`
3. in order to fix this we applied a fix which we thought could resolve the issue like below

             
          if [ ! -z "${MINIO_SSE_MASTER_KEY}" ]; then
                  keyval=($(echo "${MINIO_SSE_MASTER_KEY}" | tr ":" "\n"))
          
                  reencodedval=($(echo -n "${keyval[1]}"|xxd -r -p |base64))
          
                  MINIO_KMS_SECRET_KEY=("${keyval[0]}:$reencodedval")
                  export MINIO_KMS_SECRET_KEY
              fi
          
            The fix can be explained as follows
            1.take or reuse MINIO_SSE_MASTER_KEY 's value (not the secret ref of name/key)
            2. parse wire-master-key:${hex_encoded_key}  in 2 part - > 1st: wire-master-key  , 2nd: ${hex_encoded_key}
            3.for 2nd: ${hex_encoded_key}  -> decode it by hex and encoded by base64, save it in ${base64_encoded_key}
            4.combine the 2 part again: wire-master-key:${base64_encoded_key}
            5.add KMS_SECRET_KEY env var with the value of wire-master-key:${base64_encoded_key}
4. When we tried to read the data added using X version of minio
```
priyasekar@PriyaSekars-MacBook-Pro ibm-minio-operator %  mc cp --insecure sse-test/bucket/testing-encrypt.txt  ./test-downloaded-from-minio-1010.txt
����g�Vf�g� ��ծ)X�bM���B�ޡ"��(���d\�;�|...)st:9000/bucket/testing-encrypt.txt`. We encountered an internal error, please try again.: cause(readObjectStart: expect { or n, but found �, error found in #2 byte of ...| ������6XX�
priyasekar@PriyaSekars-MacBook-Pro ibm-minio-operator % mc cat --insecure sse-test/bucket/testing-encrypt.txt
����g�Vf�g� ��ծ)X�bM���B�ޡ"��(���d\�;�|...).cket/testing-encrypt.txt`. We encountered an internal error, please try again.: cause(readObjectStart: expect { or n, but found �, error found in #2 byte of ...| ������6XX�

```
## Configuration
We have container based environment which runs on ubi image.
SSE_S3 is used 

This is the KMS secret key thats used in both versions of minio
```
docker_kms_secret_encryption_env() {
    if [ -f "$MINIO_KMS_SECRET_KEY_FILE" ]; then
        KMS_SECRET_KEY_FILE="$MINIO_KMS_SECRET_KEY_FILE"
    else
        KMS_SECRET_KEY_FILE="/run/secrets/$MINIO_KMS_SECRET_KEY_FILE"
    fi

    if [ -f "$KMS_SECRET_KEY_FILE" ]; then
        MINIO_KMS_SECRET_KEY="$(cat "$KMS_SECRET_KEY_FILE")"
        export MINIO_KMS_SECRET_KEY
    fi
}
```

**Comment by klauspost** [comment]
@gughappriya Please specify your configuration.

**Comment by gughappriya** [comment]
> @gughappriya Please specify your configuration.

What configuration you need, can you be specific

**Comment by harshavardhana** [comment]
All your server ENV values @gughappriya

**Comment by harshavardhana** [comment]
Moved this as a discussion more appropriate. 

> if we use sse enabled in both versions of minio , when data thats added using RELEASE.2021-07-08T01-15-01Z upgraded to RELEASE.2021-11-03T03-36-36Z , results in error

Please upgrade to the latest release.

**Reply by aaronmcohen** [reply]
Good afternoon @harshavardhana, What in the latest release fixed this issue?

**Reply by harshavardhana** [reply]
> Good afternoon @harshavardhana, What in the latest release fixed this issue?

Nothing really - community users should always upgrade to the latest before reporting issues.

**Reply by gughappriya** [reply]
**Added data to bucket using old version of minio**
```
priyasekar@priyasekars-mbp ~ % mc --insecure config host add sse-test https://localhost:9000
Enter Access Key: 250efd17a20670cf778bf52128118c34a74a
Enter Secret Key:
Added `sse-test` successfully.
priyasekar@priyasekars-mbp ibm-minio-operator % mc cp --insecure --encrypt sse-test/bucket/ ./testing.txt sse-test/bucket/testing-encrypt.txt
./testing.txt:                             26 B / 26 B  
priyasekar@priyasekars-mbp ibm-minio-operator % mc stat --insecure sse-test/bucket/testing-encrypt.txt
Name      : testing-encrypt.txt
Date      : 2022-02-15 15:00:39 EST
Size      : 26 B
ETag      : 658de63034a30882f226d853ebe53612
Type      : file
Metadata  :
  Content-Type: text/plain
Encrypted :
  X-Amz-Server-Side-Encryption: AES256

priyasekar@priyasekars-mbp ibm-minio-operator %  mc cat --insecure sse-test/bucket/testing-encrypt.txt
hello welcome to our test
```
**Upgraded to latest minio version thats available now**
```
priyasekar@priyasekars-mbp ibm-minio-operator % mc cp --insecure sse-test/bucket/testing-encrypt.txt  ./test-downloaded-from-minio-old.txt
mc: <ERROR> Failed to copy `https://localhost:9000/bucket/testing-encrypt.txt`. Server side encryption specified but KMS is not configured
```

**Comment by gughappriya** [comment]
There is a slack conversation going on with minio community as well
```
[Gughappriya G](https://app.slack.com/team/U031HG1GR39)  [2 days ago](https://minio.slack.com/archives/C3NDUB8UA/p1644334583948799)
Hello everyone we are using Minio release RELEASE.2021-07-08T01-15-01Z. How to address security vulnerabilities with this release. Can someone let me know. We do not want to upgrade.

[Prakash S](https://app.slack.com/team/USMET6WLA)  [2 days ago](https://minio.slack.com/archives/C3NDUB8UA/p1644334756040119?thread_ts=1644334583.948799&cid=C3NDUB8UA)
[min.io/pricing](http://min.io/pricing) subscription plan might provide the expected options. please have a look .
as security fixes are exclusively available via SUBNET portal for subscribed customers.

[Eco](https://app.slack.com/team/UBJDUDDDY)  [2 days ago](https://minio.slack.com/archives/C3NDUB8UA/p1644348042658769?thread_ts=1644334583.948799&cid=C3NDUB8UA)
[@Gughappriya G](https://minio.slack.com/team/U031HG1GR39) as Prakash mentioned, if you have a subscription then backports can be provided for up to a year for standard, or 5 years for enterprise.  But, generally it makes more sense to upgrade releases going back that far, is there some reason you do not want to upgrade?

[Gughappriya G](https://app.slack.com/team/U031HG1GR39)  [8 hours ago](https://minio.slack.com/archives/C3NDUB8UA/p1644505748140909?thread_ts=1644334583.948799&cid=C3NDUB8UA)
yes there are issues when we upgrade to latest like the data entered using this old release was not readable with new release

[Prakash S](https://app.slack.com/team/USMET6WLA)  [8 hours ago](https://minio.slack.com/archives/C3NDUB8UA/p1644506142211659?thread_ts=1644334583.948799&cid=C3NDUB8UA)
how much of data (TiB)  do you have and what exactly is the issue and the kind of setup is used ( bare metal, k8s...)

[Eco](https://app.slack.com/team/UBJDUDDDY)  [6 hours ago](https://minio.slack.com/archives/C3NDUB8UA/p1644513100430579?thread_ts=1644334583.948799&cid=C3NDUB8UA)
yes there are issues when we upgrade to latest like the data entered using this old release was not readable with new release
I dont think this is possible if you have not manually modified something, in addition to what [@Prakash S](https://minio.slack.com/team/USMET6WLA) requested, what version did you start on and upgrade to?

[Gughappriya G](https://app.slack.com/team/U031HG1GR39)  [5 hours ago](https://minio.slack.com/archives/C3NDUB8UA/p1644516883038149?thread_ts=1644334583.948799&cid=C3NDUB8UA)
https://github.com/minio/minio/discussions/14263

GitHubGitHub
[Unable to migrate from RELEASE.2021-07-08T01-15-01Z to RELEASE.2021-11-03T03-36-36Z · Discussion #14263 · minio/minio](https://github.com/minio/minio/discussions/14263)
Expected Behavior if we use sse enabled in both versions of minio , when data thats added using RELEASE.2021-07-08T01-15-01Z upgraded to RELEASE.2021-11-03T03-36-36Z . The data should be upgraded w... (98 kB)
https://github.com/minio/minio/discussions/14263[](https://github.com/minio/minio/discussions/14263)


[Gughappriya G](https://app.slack.com/team/U031HG1GR39)  [5 hours ago](https://minio.slack.com/archives/C3NDUB8UA/p1644516894019169?thread_ts=1644334583.948799&cid=C3NDUB8UA)
Here is the detailed issue that i have raised

[Eco](https://app.slack.com/team/UBJDUDDDY)  [5 hours ago](https://minio.slack.com/archives/C3NDUB8UA/p1644517266678099?thread_ts=1644334583.948799&cid=C3NDUB8UA)
I don't see any details at all about configuration, so it's hard to pinpoint any problem without knowing things like what the server configuration looks like, how was sse-c configured etc, what other environment variables you have configured etc

[Gughappriya G](https://app.slack.com/team/U031HG1GR39)  [3 hours ago](https://minio.slack.com/archives/C3NDUB8UA/p1644524978127419?thread_ts=1644334583.948799&cid=C3NDUB8UA)
MINIO_ACCESS_KEY
MINIO_SECRET_KEY
MINIO_SSE_MASTER_KEY = wire-master-key:XYZ
are the environment variables that are set 
(edited)

[Gughappriya G](https://app.slack.com/team/U031HG1GR39)  [< 1 minute ago](https://minio.slack.com/archives/C3NDUB8UA/p1644534262776629?thread_ts=1644334583.948799&cid=C3NDUB8UA)
So to give the full story of what we tried lets call this old release(RELEASE.2021-07-08T01-15-01Z ) as X and the later release(RELEASE.2021-11-03T03-36-36Z) we tried as Y
Scenario : We added data using the X version of minio and then later upgraded to Y version. The data we added was not readable anymore
With SSE enabled we added data to buckets using X version of minio.
mc --insecure config host add sse-test https://localhost:9000/
mc ls --insecure sse-test

mc cp --insecure --encrypt sse-test/bucket/ ./testing.txt sse-test/bucket/testing-encrypt.txt
./testing.txt:                         21 B / 21 B

mc stat --insecure sse-test/bucket/testing-encrypt.txt
Name      : testing-encrypt.txt
Date      : 2021-12-13 17:53:35 CST
Size      : 21 B
ETag      : 3a5eba4d91f3dbbe047f24ce02cd2530
Type      : file
Metadata  :
  Content-Type: text/plain
Encrypted :
  X-Amz-Server-Side-Encryption: AES256

mc cat --insecure sse-test/bucket/testing-encrypt.txt
Hello Hello Hello!!!
2. When we upgraded to Y version of minio we got the following error on reading
cause: com.amazonaws.services.s3.model.AmazonS3Exception: Server side encryption specified but KMS is not configured (Service: Amazon S3; Status Code: 501; Error Code: NotImplemented; Request ID: 16BC9A97099E0EE9; S3 Extended Request ID: 52bff882-da55-46c7-a8bb-a7cb76534d2d), S3 Extended Request ID: 52bff882-da55-46c7-a8bb-a7cb76534d2d)
3. in order to fix this we applied a fix which we thought could resolve the issue like below
    if [ ! -z "${MINIO_SSE_MASTER_KEY}" ]; then
        keyval=($(echo "${MINIO_SSE_MASTER_KEY}" | tr ":" "\n"))

        reencodedval=($(echo -n "${keyval[1]}"|xxd -r -p |base64))

        MINIO_KMS_SECRET_KEY=("${keyval[0]}:$reencodedval")
        export MINIO_KMS_SECRET_KEY
    fi
The fix can be explained as follows
1.take or reuse MINIO_SSE_MASTER_KEY 's value (not the secret ref of name/key)
2. parse wire-master-key:${hex_encoded_key}  in 2 part - > 1st: wire-master-key  , 2nd: ${hex_encoded_key}
3.for 2nd: ${hex_encoded_key}  -> decode it by hex and encoded by base64, save it in ${base64_encoded_key}
4.combine the 2 part again: wire-master-key:${base64_encoded_key}
5.add KMS_SECRET_KEY env var with the value of wire-master-key:${base64_encoded_key}
4. When we tried to read the data added using X version of minio
priyasekar@PriyaSekars-MacBook-Pro ibm-minio-operator %  mc cp --insecure sse-test/bucket/testing-encrypt.txt  ./test-downloaded-from-minio-1010.txt
����g�Vf�g� ��ծ)X�bM���B�ޡ"��(���d\�;�|...)st:9000/bucket/testing-encrypt.txt`. We encountered an internal error, please try again.: cause(readObjectStart: expect { or n, but found �, error found in #2 byte of ...| ������6XX�
priyasekar@PriyaSekars-MacBook-Pro ibm-minio-operator % mc cat --insecure sse-test/bucket/testing-encrypt.txt
����g�Vf�g� ��ծ)X�bM���B�ޡ"��(���d\�;�|...).cket/testing-encrypt.txt`. We encountered an internal error, please try again.: cause(readObjectStart: expect { or n, but found �, error found in #2 byte of ...| ������6XX�
```

**Comment by gughappriya** [comment]
Slack conversation link https://minio.slack.com/archives/C3NDUB8UA/p1644334583948799

**Comment by SteveW94** [comment]
We experience the same issue when trying to upgrade our minio setup to anything later then the April 2021 release.

We also introduced the new Env-Variable MINIO_KMS_SECRET_KEY and gave them the value of the past MINIO_SSE_MASTER_KEY variable, but in base64 instead of HEX format.

The result is same as here that after setting the key, the old minios files cannot read by minio in any way as they seem to contain byte-garbage...

Our assumption is that it may be possible that those files from the old minio are not considered encrypted on the new minio and so the key (although its set) is not apllied to them in any way.

We tried a lot of stuff to desperatly update our setup to the most recent version of minio.

Is there ANY workaround/fix/migration process to this yet? I found nothing in the docu on this, but it seems an issue though...
Maybe the files need some manual "I am encrypted" tag by force?

We tried to set object tags in the new minio but it refused to set them as it was not ale to read something understandable out of the file.

Any kind of help would be much appriciated :)

Thanks

Stefan

**Comment by pruiz** [comment]
Hi,

I've faced the same issue and as 'mc mirror' was not a viable solution, I just made a fork against RELEASE.2023-01-06T18-11-18Z reintroducing support for MINIO_KMS_MASTER_KEY so we can upgrade by now and decide what to do next. Maybe this is of any use for anyone else within the community, so there is my fork.

Also, I will probably re-implement this support as an extension to KES (or something supporting KES API), so in the future we can go back to using official minio binaries, and have an upgrade path for key rotation.

Regards
Pablo

**Reply by SteveW94** [reply]
That would be super-awesome to have some kind of upgrade-path or even use the master key as-is again.
Looking much forward to have that in the official binaries again, as so far beeing also stuck on the 2021 version^^

Thanks a lot! :)

**Reply by pruiz** [reply]
Hi @SteveW94 on my fork I've re-added support for MINIO_KMS_MASTER_KEY against a current release, and provided binaries (inform of rpm packages, but the binary inside can be used on any linux) at: https://copr.fedorainfracloud.org/coprs/pruiz/minio/build/5219338/

**Reply by SteveW94** [reply]
Is there any chance to have this also as a windows binary?

**Reply by pruiz** [reply]
Hi @SteveW94,

I've uploaded binaries to https://github.com/pruiz/minio/releases/tag/RELEASE.2023-01-06T18-11-18Z.hotfix.95dbebd49

However windows binaries are untested, I just built it using GOOS=windows, hope it works fine.

Regards
Pablo

**Reply by SteveW94** [reply]
Wow thanks :)

I ll try it as soon as our project allows it.

I hope the official binaries somehow will include some migration possibility or sth like that at least which does not mean youn have to first decrypt everything with the old version before migrating to the new one^^
